### PR TITLE
Command rewrite

### PR DESCRIPTION
**Pull request by AzatothTheAwakened**
Started to rewrite the command call structure such as that the main method is a lot smaller.
Now every Class needs to just inherit from one baseclass to be added as command instead of previously having to add them all manually. An example of this can be found in the file `src/main/java/EventsAndCommands/FunEvents/GoodBotEvent.java`.

### CODE REVIEW

File: src/main/java/EventsAndCommands/ChatEventDistributor.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * CommandHandler
+ */
+public class ChatEventDistributor extends ListenerAdapter{
+
+    private static ChatEventDistributor instance = null;
+
+    private Set<ChatEventHandler> eventHandlers;
+
+    public static ChatEventDistributor getInstance(){
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventDistributor.java]: This is not thread safe, but JDA is quite multi threaded. Not sure where it is used, but please add *at least* a doc comment.

File: src/main/java/EventsAndCommands/ChatEventDistributor.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import net.dv8tion.jda.core.hooks.ListenerAdapter;
+
+import org.reflections.*;
+
+/**
+ * CommandHandler
+ */
+public class ChatEventDistributor extends ListenerAdapter{
+
+    private static ChatEventDistributor instance = null;
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventDistributor.java]: Why do you explicitly init it?

File: src/main/java/EventsAndCommands/ChatEventDistributor.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        Reflections reflections = new Reflections(this.getClass().getPackageName());
+
+        var handlers = reflections.getSubTypesOf(c);
+
+        for (var handler : handlers) {
+            try{
+                Constructor<?> ctor = handler.getConstructor();
+                ctor.newInstance();
+            } catch (Exception e){
+                // TODO: Handle
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventDistributor.java]: At least log it in the meantime?

File: src/main/java/EventsAndCommands/ChatEventDistributor.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    public void registerChatEventHandler(ChatEventHandler handler){
+        eventHandlers.add(handler);
+    }
+
+    public void unregisterChatEventHandler(ChatEventHandler handler){
+        eventHandlers.remove(handler);
+    }
+
+    public void load(Class<?> c){
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventDistributor.java]: What does this method do? It has no return type and does not add anything to a field.

Does it rely on side effects in the constructors? That sounds adventurous.

File: src/main/java/EventsAndCommands/ChatEventHandler.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    protected void register(){
+        ChatEventDistributor.getInstance().registerChatEventHandler(this);
+    }
+
+    protected void unregister(){
+        ChatEventDistributor.getInstance().unregisterChatEventHandler(this);
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventHandler.java]: Uff. Why is this included? Might want to add some javadoc that justifies using the finalize machinery.

File: src/main/java/EventsAndCommands/ChatEventHandler.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    public abstract void receiveCommand(GuildMessageReceivedEvent event);
+
+    /**
+     * Determines if an event would trigger the command. If the given event evaluates to true, 
+     * {@see ChatEventHandler#receiveCommand(GuildMessageReceivedEvent)} gets called.
+     * @param event The event that might trigger the command.
+     * @return true if the the command is triggered, false if not.
+     */
+    public abstract boolean trigger(GuildMessageReceivedEvent event);
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventHandler.java]: `trigger` sounds like it *triggers* it. Consider something like `shouldTrigger`, `triggers`, `isTriggeredBy` or something similar.

File: src/main/java/EventsAndCommands/ChatEventHandler.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    public ChatEventHandler(){
+        register();
+    }
+    /**
+     * This method is executed when the command got triggered trough {@see ChatEventHandler#trigger(GuildMessageReceivedEvent)}.
+     * This method should only handle what happens after the command was triggered, i.e. responses or whatever,
+     * @param event the message that triggered the event.
+     */
+    public abstract void receiveCommand(GuildMessageReceivedEvent event);
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventHandler.java]: Does this method execute a command If so `execute` might be nicer. If not `onCommand` might fit better? Not sure.

File: src/main/java/EventsAndCommands/FunEvents/GoodBotEvent.java [hunk 8]
```diff
@@ -7,48 +7,50 @@
...
+                        .queue();
+            }
 
+        });
+    }
 
+    @Override
+    public boolean trigger(GuildMessageReceivedEvent event) {
+        String msg = event.getMessage().getContentRaw().toLowerCase();
+        boolean r = (msg.contains("good bot") || msg.contains("bad bot"));
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/FunEvents/GoodBotEvent.java]: You don't need the local variable here, you can just return the boolean expression.

File: src/main/java/EventsAndCommands/ChatEventHandler.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * This class provides the base structure of a chatevent that might be handled. 
+ * To implement a new ChatEventHandler it is sufficient to just inherit from this 
+ * Class and override {@see ChatEventHandler#trigger(GuildMessageReceivedEvent)}
+ * and {@see ChatEventHandler#receiveCommand(GuildMessageReceivedEvent)}.
+ */
+public abstract class ChatEventHandler {
+
+    public ChatEventHandler(){
+        register();
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventHandler.java]: Having this kind of side-effect in a constructor sounds really dangerous. I'd prefer to make the (un) register methods public and let the auto-discovery handle it.

Maybe adjust the discovery to:
```java
loadAllClasses(Class<T> superType, Consumer<T> initialization)
```

File: src/main/java/EventsAndCommands/ChatEventDistributor.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        Reflections reflections = new Reflections(this.getClass().getPackageName());
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventDistributor.java]: I'd add that it just looks *in this classe's package* to the javadoc.

File: src/main/java/EventsAndCommands/ChatEventDistributor.java (hunk 1 above)
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/ChatEventDistributor.java]: This is not thread safe, but JDA is quite multi threaded. Not sure where it is used, but please add at least a doc comment.

And I am not sure a singleton is a good fit. If you pass an initialization consumer to the `load` method, you can make this a normal class. This works as you can create one instance in the main class and just register it in the initialization consumer. This also removes the dependency of the listener to the manager. If you want to keep it, you can inject the instance in the initialization block.

**Review by I-Al-Istannen — COMMENTED** [review]
Highlighted some small parts that could *maybe* benefit from some further pondering.

**Comment by AzatothTheAwakened** [comment]
Removed the first approach of dynamically loading classes that inherit a specific abstract class. Now every class that inherits from Command and ListenerAdaptor are candidates for being hotloaded. 
Needed Databases and other dependent instances are injected as soon as a suitable ctor is found.
still needs some cleanup tho

File: src/main/java/EventsAndCommands/ChatEventDistributor.java (hunk 1 above)
Comment (AzatothTheAwakened) [reply src/main/java/EventsAndCommands/ChatEventDistributor.java]: Has ben resolved since the distribution of events to the fitting handlers is now handled by JDA itself again

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        setConfig(config);
+        setWaiter(waiter);
+    }
+
+    public void loadCommands() {
+        Reflections reflections = new Reflections(this.getClass().getPackageName());
+
+        Set<Class<? extends Command>> cmds = reflections.getSubTypesOf(Command.class);
+        cmds = cmds.stream()
+            .filter(x -> x.getAnnotation(IgnoreCommand.class) == null)
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: There is a Class#isAnnotationPresent method.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+        setDb(db);
+        setConfig(config);
+        setWaiter(waiter);
+    }
+
+    public void loadCommands() {
+        Reflections reflections = new Reflections(this.getClass().getPackageName());
+
+        Set<Class<? extends Command>> cmds = reflections.getSubTypesOf(Command.class);
+        cmds = cmds.stream()
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: I'd directly stream and filter instead of reassigning the collection.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+            List<Constructor<?>> ctors = Arrays.asList(cmd.getConstructors());
+            for (var ctor : ctors) {
+                if (hasAcceptableParams(ctor)) {
+                    try {
+                        instances.add(injectCommand(ctor));
+                    } catch (InvocationTargetException e){
+                        e.getCause().printStackTrace();
+                    } catch (Exception e) {
+                        // TODO: Log
+                        System.err.println(e.getMessage());
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: Print the stacktrace, so you can at least see where it originated.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+        List<Command> instances = new ArrayList<>();
+
+        for (var cmd : cmds) {
+            List<Constructor<?>> ctors = Arrays.asList(cmd.getConstructors());
+            for (var ctor : ctors) {
+                if (hasAcceptableParams(ctor)) {
+                    try {
+                        instances.add(injectCommand(ctor));
+                    } catch (InvocationTargetException e){
+                        e.getCause().printStackTrace();
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: Why do you only print the cause? It probably has a good reason, but what is it?

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+                } else {
+                    System.err.println("No suitable constructor for " + cmd.getName());
+                }
+            }
+        }
+
+        setLoadedCommands(instances);
+    }
+
+    public void loadListenerAdaptor() {
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: You can abstract the reflections+subtypes+filtering+collecting out in its own method.

Likely you can even extend that to deduplicate the two load methods into one.
*Might* be useful, even though you just use that pattern twice.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+        setLoadedListeners(instances);
+    }
+
+    /**
+     * 
+     * @param ctors
+     * @return
+     */
+    public ListenerAdapter injectListenerAdaptor(Constructor<?> ctor) throws Exception{
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: Expect for the final cast this looks identical to injectCommand? I'd use one more generic method, maybe casting to a type token class you pass it.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+        acceptableParams.add(EventWaiter.class);
+
+        return acceptableParams.containsAll(Arrays.asList(ctor.getParameterTypes()));
+    }
+
+    /**
+     * @return the eventHandlers
+     */
+    public List<Command> getLoadedCommands() {
+        return loadedCommands;
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: You likely want to return an unmodifiable wrapper (Collections.unmodifiableList) **if** modifications to your inner `loadeCommands` are not wanted.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    /**
+     * @param eventHandlers the eventHandlers to set
+     */
+    public void setLoadedCommands(List<Command> loadedCommands) {
+        this.loadedCommands = loadedCommands;
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: I'd make a defensive copy, if modifications to the parameter should not write through to your underlying loadedCommands list.

File: src/main/java/EventsAndCommands/IgnoreCommand.java [hunk 19]
```diff
@@ -0,0 +1,13 @@
+package EventsAndCommands;
+
+import java.lang.annotation.*;
+
+/**
+ * IgnoreCommand
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/IgnoreCommand.java]: "Commands annotated with this annotation will *not automatically be instantiated and registered*"?

**Review by I-Al-Istannen — COMMENTED** [review]
Look a lot nicer :)

File: src/main/java/EventsAndCommands/CommandLoader.java (hunk 14 above)
Comment (AzatothTheAwakened) [reply src/main/java/EventsAndCommands/CommandLoader.java]: InvocationTargetException wraps another Exception that was thrown inside the code that got invoked via reflection. In this case getCause() returns the actually thrown exception and not the wrapper.

See: <https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/lang/reflect/InvocationTargetException.html#getCause()>

Comment (I-Al-Istannen) [reply src/main/java/EventsAndCommands/CommandLoader.java]: I am aware, but then you lose the context where exactly the reflective call was:
```bash
$ javac Test.java && java Test
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at Test.main(Test.java:9)
Caused by: java.lang.RuntimeException: Hey
	at Test.test(Test.java:4)
	... 5 more

# Changed to getCause().printStackTrace()

$ javac Test.java && java Test
java.lang.RuntimeException: Hey
	at Test.test(Test.java:4)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at Test.main(Test.java:9)

```

Comment (AzatothTheAwakened) [reply src/main/java/EventsAndCommands/CommandLoader.java]: Fixed it, didn'T notice that it still prints the wrapped cause since my IDE wraps stacktraces

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 20]
```diff
@@ -2,128 +2,95 @@
...
-import com.mongodb.client.MongoDatabase;
 import com.jagrosh.jdautilities.commons.waiter.EventWaiter;
+import com.mongodb.client.MongoDatabase;
+
+import org.reflections.Reflections;
 
 /**
- * CommandHandler
+ * Autodetects commands that implement a generic type T. Can be used to
+ * dynaically load up all commands without explicitly instantiating them.
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: `dynaically`

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 21]
```diff
@@ -2,128 +2,95 @@
...
+ * Injects needed dependencies into the instances individually. T = The
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: should be an `@param <T> The superclass whose subclasses should be loaded`

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 22]
```diff
@@ -2,128 +2,95 @@
...
+    public CommandLoader(Class<T> token, MongoDatabase db, Properties config, EventWaiter waiter) {
         setDb(db);
         setConfig(config);
         setWaiter(waiter);
+        setToken(token);
     }
 
-    public void loadCommands() {
+    /**
+     * Dynamically instanciates all Classes that inherit from type T. But ignores classes that are annotated with {@see IgnoreCommand}
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: Shouldn't this be an `@link`?

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 23]
```diff
@@ -2,128 +2,95 @@
...
+                if (instance != null) {
+                    instances.add(instance);
                 } else {
-                    System.err.println("No suitable constructor for " + cmd.getName());
+                    System.err.println("No suitable constructor found for " + cmd.getName());
                 }
+            } catch (InvocationTargetException e) {
+                e.printStackTrace();
+            } catch (Exception e) {
+                System.err.println(e.getMessage());
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: This is already included in the stacktrace.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 24]
```diff
@@ -2,128 +2,95 @@
...
+     * @param cl A reflective Class that needs to be instanced.
+     * @return A concrete object derived from cl. Or null if no suitable constructor
+     *         was found.
+     * @throws Exception gets thrown if something goes wrong with reflective
+     *                   operations.
      */
-    public ListenerAdapter injectListenerAdaptor(Constructor<?> ctor) throws Exception{
-        Class<?>[] paramTypes = ctor.getParameterTypes();
-        Object[] parameters = new Object[paramTypes.length];
+    private T inject(Class<? extends T> cl) throws Exception {
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: Use `ReflectiveOperationException` then

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 25]
```diff
@@ -135,30 +102,42 @@ public Command injectCommand(Constructor<?> ctor) throws Exception{
             }
         }
 
-        return (Command) ctor.newInstance(parameters);
+        return (T) ctor.newInstance(parameters);
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: `cl.cast(ctor.newInstance(parameters))` might look slightly nicer. Not sure.

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 26]
```diff
@@ -135,30 +102,42 @@ public Command injectCommand(Constructor<?> ctor) throws Exception{
...
+    private Constructor<? extends T> findFittingConstructor(Class<? extends T> cl) {
         final Set<Class<?>> acceptableParams = new HashSet<>();
         acceptableParams.add(MongoDatabase.class);
         acceptableParams.add(Properties.class);
         acceptableParams.add(EventWaiter.class);
 
-        return acceptableParams.containsAll(Arrays.asList(ctor.getParameterTypes()));
-    }
+        List<Constructor<?>> ctors = Arrays.asList(cl.getConstructors());
+        ctors.sort((x, y) -> x.getParameterCount() - y.getParameterCount());
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: `-` is a bit dangerous for comparators. I'd prefer
```java
ctors.sort(Comparator.comparingInt(Constructor::getParameterCount));

// or for the reverse order
ctors.sort(comparing(Constructor::getParameterCount, reverseOrder()))
```

File: src/main/java/EventsAndCommands/CommandLoader.java [hunk 27]
```diff
@@ -135,30 +102,42 @@ public Command injectCommand(Constructor<?> ctor) throws Exception{
...
-     * @param eventHandlers the eventHandlers to set
-     */
-    public void setLoadedCommands(List<Command> loadedCommands) {
-        this.loadedCommands = loadedCommands;
+        for (var ctor : ctors) {
+            Class<?>[] params = ctor.getParameterTypes();
+            if (acceptableParams.containsAll(Arrays.asList(ctor.getParameterTypes()))) {
+                try {
+                    return cl.getConstructor(params);
+                } catch (NoSuchMethodException e) {
```
Comment (I-Al-Istannen) [inline src/main/java/EventsAndCommands/CommandLoader.java]: Is this github or is that really a syntax error here? Where is the closing parenthesis for that catch?

EDIT: Oh. If you want to swallow it, rename `e` to `ignored` or sth that conveys your intent a bit better.

**Review by I-Al-Istannen — APPROVED** [review]
Looks nice

File: src/main/java/EventsAndCommands/CommandLoader.java (hunk 21 above)
Comment (AzatothTheAwakened) [reply src/main/java/EventsAndCommands/CommandLoader.java]: param <T>? I am unaware of that you use param to annotate generic types

Comment (I-Al-Istannen) [reply src/main/java/EventsAndCommands/CommandLoader.java]: You do :)
grafik
